File: src/products/s3-bucket/s3-bucket.template.ts
import * as cdk from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';
import { CfnParameter, CfnOutput, Fn } from '@aws-cdk/core';
import { ProductTemplate } from '../product.model';

class S3StackTemplate extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const bucketName = new CfnParameter(this, 'BucketName');

    this.templateOptions.metadata = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: [
          {
            Label: { default: 'Bucket Configuration' },
            Parameters: [bucketName.logicalId]
          }
        ],
        ParameterLabels: {
          [bucketName.logicalId]: {
            default: 'Which name should the bucket have'
          }
        }
      }
    }

    const bucket = new Bucket(this, 'test-bucket', {
      bucketName: bucketName.valueAsString
    });

    new CfnOutput(this, 'S3Id', {
      value: bucket.bucketArn,
      exportName: Fn.sub('${AWS::StackName}-S3Id')
    })
  }
}

export const S3ProductTemplate: ProductTemplate = {
  stack: S3StackTemplate,
  name: 'S3Product'
};

File: python/src/products/s3_bucket/s3_bucket_template.py
#!/usr/bin/env python3
from aws_cdk import core as cdk
from aws_cdk.aws_s3 import Bucket
from ..product_model import ProductTemplate
from aws_cdk.core import CfnParameter, CfnOutput, Fn


class S3StackTemplate(cdk.Stack):
    def __init__(self, scope: cdk.Construct, id: str, **kwargs):
        super().__init__(scope, id, **kwargs)
        bucketName = CfnParameter(self, "BucketName")
        self.template_options.metadata = {
            'AWS::CloudFormation::Interface': {
                'ParameterGroups': [
                    {
                        'Label': {'default': 'Bucket Configuration'},
                        'Parameters': [bucketName.logical_id]
                    }
                ],
                'ParameterLabels': {
                    bucketName.logical_id: {
                        'default': 'Which name should the bucket have'
                    }
                }
            }
        }

        bucket = Bucket(self, 'test-bucket',
                        bucket_name=bucketName.value_as_string)

        CfnOutput(self, 'S3Id', value=bucket.bucket_arn,
                  export_name=Fn.sub('${AWS::StackName}-S3Id'))


S3ProductTemplate = ProductTemplate(
    stack=S3StackTemplate,
    name='S3Product'
)
